### ISSUE TITLE
请问怎么把返回的所有结果合并起来？

### ISSUE THREAD

**Issue by Maxhyl** [firslov/llama2-api#2]
我修改了streamllama.py文件chat_completion，执行完for循环后直接返回结果，然后直接return结果，但是这样第一次调用是没问题，第二次之后就会超时失败了

**Comment by firslov** [comment]
> 我修改了streamllama.py文件chat_completion，执行完for循环后直接返回结果，然后直接return结果，但是这样第一次调用是没问题，第二次之后就会超时失败了

能否说明下代码修改细节和报错信息

**Comment by Maxhyl** [comment]
修改streamllama.py文件的generate方法如下：
`
def generate(
        self,
        prompt_tokens: List[List[int]],
        max_gen_len: int,
        temperature: float = 0.6,
        top_p: float = 0.9,
    ):
        params = self.model.params
        bsz = len(prompt_tokens)
        assert bsz <= params.max_batch_size, (bsz, params.max_batch_size)

        min_prompt_len = min(len(t) for t in prompt_tokens)
        max_prompt_len = max(len(t) for t in prompt_tokens)
        assert max_prompt_len <= params.max_seq_len
        total_len = min(params.max_seq_len, max_gen_len + max_prompt_len)

        pad_id = self.tokenizer.pad_id
        tokens = torch.full((bsz, total_len), pad_id, dtype=torch.long, device="cuda")
        for k, t in enumerate(prompt_tokens):
            tokens[k, : len(t)] = torch.tensor(t, dtype=torch.long, device="cuda")

        prev_pos = 0
        sentence = ""
        sentence_list = []
        input_text_mask = tokens != pad_id
        res = ""
        for cur_pos in range(min_prompt_len, total_len):
            logits = self.model.forward(tokens[:, prev_pos:cur_pos], prev_pos)

            if temperature > 0:
                probs = torch.softmax(logits[:, -1] / temperature, dim=-1)
                next_token = sample_top_p(probs, top_p)
            else:
                next_token = torch.argmax(logits[:, -1], dim=-1)

            next_token = next_token.reshape(-1)
            # only replace token if prompt has already been generated
            next_token = torch.where(
                input_text_mask[:, cur_pos], tokens[:, cur_pos], next_token
            )
            tokens[:, cur_pos] = next_token
            prev_pos = cur_pos

            sentence_list.append(next_token.item())
            new_sentence = self.tokenizer.decode(sentence_list)
            delta = new_sentence.replace(sentence, "")
            sentence = new_sentence
            res += delta
            # if next_token == self.tokenizer.eos_id:
            #     yield json.dumps(
            #         {
            #             "object": "chat.completion.chunk",
            #             "model": "llama2-13B-chat",
            #             "choices": [
            #                 {
            #                     "delta": {},
            #                     "index": 0,
            #                     "finish_reason": "stop",
            #                 }
            #             ],
            #         }
            #     )
            #     yield "[DONE]"
            #     break
            # else:
            #     yield json.dumps(
            #         {
            #             "object": "chat.completion.chunk",
            #             "model": "llama2-13B-chat",
            #             "choices": [
            #                 {
            #                     "delta": {"content": delta},
            #                     "index": 0,
            #                     "finish_reason": None,
            #                 }
            #             ],
            #         }
            #     )
        return {"response": res}
`

api.py文件的return修改为：return EventSourceResponse(my_generator(g))修改为return g

**Comment by firslov** [comment]
OK，首先返回的数据要进行格式化，哪怕是一次性返回所有数据，可以参考我代码里的形式；其次如果使用第三方的前端可能默认采用流式输出，需要最后返回一个“stop”，才能结束一次socket通信。

另外，llama2原生api很不友好，建议可以尝试下baichuan2-7b/baichuan2-13b。

**Comment by firslov** [comment]
哦，我知道了，你修改代码后影响了双卡进程的同步，第一个办法是修改`api.py`中`rank1`的`dist.broadcast_object_list()`相关代码，第二个办法是不修改源代码中的返回值，使用支持流式输出的前端工具，比如：https://github.com/Bin-Huang/chatbox

**Comment by Maxhyl** [comment]
好的，谢谢，我按照您的建议先尝试下
